File: backends/go-gin/tests/gate_test.go
package tests

import (
	"fmt"
	"net/http"
	"testing"
)

func TestGate_Issue(t *testing.T) {
	truncateTables(t)

	t.Run("JWTが発行できる", func(t *testing.T) {
		c := createClient(t, nil)
		w := do(http.MethodGet, "/api/gate/issue?member=member-001", nil,
			withBearer(c.AccessToken))
		if w.Code != http.StatusOK {
			t.Errorf("want 200, got %d: %s", w.Code, w.Body.String())
		}
		body := parseBody(w)
		if body["token"] == nil {
			t.Error("token not found in response")
		}
	})

	t.Run("memberパラメーター未指定で400が返る", func(t *testing.T) {
		c := createClient(t, map[string]interface{}{"identifier": "c-for-member-test", "email": "mt@example.com"})
		w := do(http.MethodGet, "/api/gate/issue", nil,
			withBearer(c.AccessToken))
		if w.Code != http.StatusBadRequest {
			t.Errorf("want 400, got %d", w.Code)
		}
	})

	t.Run("無効なトークンで401が返る", func(t *testing.T) {
		w := do(http.MethodGet, "/api/gate/issue?member=member-001", nil,
			withBearer("invalid-token"))
		if w.Code != http.StatusUnauthorized {
			t.Errorf("want 401, got %d", w.Code)
		}
	})

	t.Run("Bearerトークンなしで401が返る", func(t *testing.T) {
		w := do(http.MethodGet, "/api/gate/issue?member=member-001", nil)
		if w.Code != http.StatusUnauthorized {
			t.Errorf("want 401, got %d", w.Code)
		}
	})
}

func TestGate_Verify(t *testing.T) {
	truncateTables(t)

	t.Run("JWTが検証できる", func(t *testing.T) {
		c := createClient(t, nil)

		// JWT を発行
		issueW := do(http.MethodGet, "/api/gate/issue?member=member-001", nil,
			withBearer(c.AccessToken))
		if issueW.Code != http.StatusOK {
			t.Fatalf("issue failed: %d %s", issueW.Code, issueW.Body.String())
		}
		jwtToken := parseBody(issueW)["token"].(string)

		// JWT を検証
		verifyURL := fmt.Sprintf("/api/gate/client/%s/verify?token=%s", c.Identifier, jwtToken)
		w := do(http.MethodGet, verifyURL, nil)
		if w.Code != http.StatusOK {
			t.Errorf("want 200, got %d: %s", w.Code, w.Body.String())
		}
	})

	t.Run("tokenパラメーター未指定で400が返る", func(t *testing.T) {
		c := createClient(t, map[string]interface{}{"identifier": "c-verify", "email": "v@example.com"})
		w := do(http.MethodGet, fmt.Sprintf("/api/gate/client/%s/verify", c.Identifier), nil)
		if w.Code != http.StatusBadRequest {
			t.Errorf("want 400, got %d", w.Code)
		}
	})

	t.Run("存在しないidentifierで403が返る", func(t *testing.T) {
		w := do(http.MethodGet, "/api/gate/client/unknown-client/verify?token=dummy", nil)
		if w.Code != http.StatusForbidden {
			t.Errorf("want 403, got %d", w.Code)
		}
	})

	t.Run("無効なJWTで401が返る", func(t *testing.T) {
		createClient(t, map[string]interface{}{"identifier": "jwt-test-client", "email": "jwt@example.com"})
		w := do(http.MethodGet, "/api/gate/client/jwt-test-client/verify?token=invalid.jwt.token", nil)
		if w.Code != http.StatusUnauthorized {
			t.Errorf("want 401, got %d", w.Code)
		}
	})
}

File: backends/ts-hono/src/test/gate.test.ts
import { describe, test, expect } from "vitest";
import { createApp } from "../app.js";
import { makeClientRecord } from "./helpers.js";

const app = createApp();

describe("Gate", () => {
  describe("GET /api/gate/issue", () => {
    test("JWTが発行できる", async () => {
      const c = await makeClientRecord();
      const res = await app.request("/api/gate/issue?member=member-001", {
        headers: { Authorization: `Bearer ${c.token}` },
      });
      expect(res.status).toBe(200);
      const body = await res.json() as Record<string, unknown>;
      expect(body.token).toBeTruthy();
    });

    test("memberパラメーター未指定で400が返る", async () => {
      const c = await makeClientRecord({ identifier: `c-no-member-${Date.now()}` });
      const res = await app.request("/api/gate/issue", {
        headers: { Authorization: `Bearer ${c.token}` },
      });
      // member は required Query なので 400
      expect(res.status).toBe(400);
    });

    test("無効なトークンで401が返る", async () => {
      const res = await app.request("/api/gate/issue?member=member-001", {
        headers: { Authorization: "Bearer invalid-token" },
      });
      expect(res.status).toBe(401);
    });
  });

  describe("GET /api/gate/client/:identifier/verify", () => {
    test("JWTが検証できる", async () => {
      const c = await makeClientRecord();
      // JWT を発行
      const issueRes = await app.request("/api/gate/issue?member=member-001", {
        headers: { Authorization: `Bearer ${c.token}` },
      });
      expect(issueRes.status).toBe(200);
      const { token: jwt } = await issueRes.json() as { token: string };

      // JWT を検証
      const res = await app.request(
        `/api/gate/client/${c.identifier}/verify?token=${jwt}`,
      );
      expect(res.status).toBe(200);
    });

    test("存在しないidentifierで401か404が返る", async () => {
      const res = await app.request("/api/gate/client/unknown-client/verify?token=dummy");
      expect([401, 404]).toContain(res.status);
    });

    test("無効なJWTで401が返る", async () => {
      const c = await makeClientRecord({ identifier: `c-invalid-jwt-${Date.now()}` });
      const res = await app.request(
        `/api/gate/client/${c.identifier}/verify?token=invalid.jwt.token`,
      );
      expect(res.status).toBe(401);
    });
  });
});
